include/igl/predicates/predicates.cpp

```diff
@@ -0,0 +1,268 @@
+// This file is part of libigl, a simple c++ geometry processing library.
+//
+// Copyright (C) 2019 Qingnan Zhou <qnzhou@gmail.com>
+//
+// This Source Code Form is subject to the terms of the Mozilla Public License 
+// v. 2.0. If a copy of the MPL was not distributed with this file, You can 
+// obtain one at http://mozilla.org/MPL/2.0/.
+#include <igl/predicates/predicates.h>
+#include <predicates.h>
+
+namespace igl {
+namespace predicates {
+
+using REAL = double;
+
+template<typename Vector2D>
+IGL_INLINE Orientation orient2d(
+        const Eigen::MatrixBase<Vector2D>& pa,
+        const Eigen::MatrixBase<Vector2D>& pb,
+        const Eigen::MatrixBase<Vector2D>& pc) {
+
+    using Point = Eigen::Matrix<REAL, Eigen::Dynamic, Eigen::Dynamic>;
+    const Point& a = pa.template cast<REAL>();
+    const Point& b = pb.template cast<REAL>();
+    const Point& c = pc.template cast<REAL>();
+
+    auto r = ::orient2d(
+            const_cast<REAL*>(a.data()),
+            const_cast<REAL*>(b.data()),
+            const_cast<REAL*>(c.data()));
+
+    if (r > 0) return Orientation::POSITIVE;
+    else if (r < 0) return Orientation::NEGATIVE;
+    else return Orientation::COLLINEAR;
+}
+
+template<typename Vector3D>
+IGL_INLINE Orientation orient3d(
+        const Eigen::MatrixBase<Vector3D>& pa,
+        const Eigen::MatrixBase<Vector3D>& pb,
+        const Eigen::MatrixBase<Vector3D>& pc,
+        const Eigen::MatrixBase<Vector3D>& pd) {
+
+    using Point = Eigen::Matrix<REAL, Eigen::Dynamic, Eigen::Dynamic>;
+    const Point& a = pa.template cast<REAL>();
```

Comment:
Since you're evaluating it, wouldn't it be better to store a `Point a = ...` and avoid the `const_cast<>`?

Reply:
When `pa`'s Scalar type is the same as `REAL`, the `cast` does not do any copying.  `Point a = ...` will force an unnecessary copy.

Reply:
If you pass a matrix expression it will still allocate a temporary.